### PR TITLE
Fix loading older fp16 optimizer checkpoints.

### PR DESCRIPTION
**Pull request by stephenroller**
**Patch description**
Sometime in the past two months, NVIDIA's APEX library changed the way it's storing/loading optimizer state from disk. As such, some models I trained in mid-february were unable to load their optimizer state successfully. This patch fixes by only loading the non-fp16 wrapped optimizer state from disk, and waiting to do the wrapping last.

This also happens to fix a bug where switching from fp16 training to fp32 training would fail to load the state from disk

**Testing steps**
- Tried loading an old model that contains the bad fp16 params
- Tried training and then resuming a checkpoint with the new fp16 params
- Tried taking the fp16 trained state and resuming in fp32 mode.

### CODE REVIEW

File: parlai/core/torch_agent.py [hunk 1]
```diff
@@ -665,15 +665,23 @@ def init_optim(self, params, optim_states=None, saved_optim_type=None):
...
-            self.optimizer = fp16_optimizer_wrapper(self.optimizer)
 
         if optim_states:
             if saved_optim_type != opt['optimizer']:
-                print('WARNING: not loading optim state since optim class '
-                      'changed.')
+                print('WARNING: not loading optim state since optim class changed.')
             else:
+                if 'loss_scaler' in optim_states:
+                    # previously, this code used to load up the fp16 optimizer
```
Comment (emilydinan) [inline parlai/core/torch_agent.py]: wowza that's one big comment

Comment (emilydinan) [reply parlai/core/torch_agent.py]: but very helpful 😄

File: parlai/core/torch_agent.py [hunk 2]
```diff
@@ -1370,8 +1381,15 @@ def state_dict(self):
...
-        if hasattr(self, 'optimizer'):  # save optimizer params
-            states['optimizer'] = self.optimizer.state_dict()
+        # save optimizer params
+        if hasattr(self, 'optimizer'):
+            # don't store the fp16 wrapper
+            if self.fp16:
+                # store the "true" optimizer, not the fp16 wrapped version
+                states['optimizer'] = self.optimizer.optimizer.state_dict()
+            else:
+                # nothing abnormal.
```
Comment (emilydinan) [inline parlai/core/torch_agent.py]: kind of a weird comment lol

**Comment by stephenroller** [comment]
This didn't work like i expected, needs further work.

Abandoning for to mark that.

**Pull request closed without merging**